pulsar-client/src/main/java/org/apache/pulsar/client/impl/ConsumerImpl.java

```diff
@@ -1907,18 +1907,17 @@ protected void completeOpBatchReceive(OpBatchReceive<T> op) {
             return CompletableFuture.completedFuture(Collections.emptyList());
         }
         List<MessageIdData> data = new ArrayList<>(messageIds.size());
```

Comment:
There are seem another potential race condition, use `ArrayList` in possible `multithread` scenario is unsafe.

Reply:
I've create #14687 to continue this discussion.

Reply:
Ah, yes.